core/src/main/scala/epollcat/unsafe/KqueueExecutorScheduler.scala

```diff
@@ -0,0 +1,170 @@
+    val timeoutIsInfinite = timeout == Duration.Inf
+    val timeoutIsZero = timeout == Duration.Zero
+    val noCallbacks = callbacks.isEmpty
+    val changeCount = changes.size()
+
+    if ((timeoutIsInfinite || timeoutIsZero) && noCallbacks && changeCount == 0)
+      false // nothing to do here. refer to scaladoc on PollingExecutorScheduler#poll
+    else {
+
+      val changelist = stackalloc[kevent64_s](changeCount.toLong)
+      var change = changelist
+      var finalChangeCount = 0
+      while (!changes.isEmpty()) {
+        val evAdd = changes.poll()
+        if (!evAdd.canceled) {
+          change.ident = evAdd.fd.toULong
+          change.filter = evAdd.filter
+          change.flags = (EV_ADD | EV_CLEAR).toUShort
+          change.udata = EventNotificationCallback.toPtr(evAdd.cb)
+          change += 1
+          finalChangeCount += 1
+        }
+      }
+
+      val timeoutSpec =
+        if (timeoutIsInfinite || timeoutIsZero) null
+        else {
+          val ts = stackalloc[time.timespec]()
+          val sec = timeout.toSeconds
+          ts.tv_sec = sec
+          ts.tv_nsec = (timeout - sec.seconds).toNanos
+          ts
+        }
+
+      val eventlist = stackalloc[kevent64_s](maxEvents.toLong)
+      val flags = (if (timeoutIsZero) KEVENT_FLAG_IMMEDIATE else KEVENT_FLAG_NONE).toUInt
+      val triggeredEvents =
+        kevent64(kqfd, changelist, finalChangeCount, eventlist, maxEvents, flags, timeoutSpec)
+
+      if (triggeredEvents >= 0) {
+        var i = 0
+        var event = eventlist
+        while (i < triggeredEvents) {
+          if ((event.flags.toLong & EV_ERROR) != 0) {
+
+            // TODO it would be interesting to propagate this failure via the callback
+            reportFailure(
+              new RuntimeException(
+                s"kevent64: flags=${event.flags.toHexString} errno=${event.data}"
+              )
+            )
+
+          } else if (callbacks.contains(event.ident.toLong)) {
+            val filter = event.filter
+            val cb = EventNotificationCallback.fromPtr(event.udata)
+
+            try {
+              cb.notifyEvents(filter == EVFILT_READ, filter == EVFILT_WRITE)
+            } catch {
+              case NonFatal(ex) =>
+                reportFailure(ex)
+            }
+          }
+
+          i += 1
+          event += 1
+        }
+      } else {
+        throw new RuntimeException(s"kevent64: ${errno.errno}")
+      }
+
+      !changes.isEmpty() || callbacks.nonEmpty
+    }
+  }
+
+  def monitor(fd: Int, reads: Boolean, writes: Boolean)(
+      cb: EventNotificationCallback): Runnable = {
+
+    val readEvent =
+      if (reads)
+        EvAdd(fd, EVFILT_READ, cb)
+      else null
+
+    val writeEvent =
+      if (writes)
+        EvAdd(fd, EVFILT_WRITE, cb)
+      else null
+
+    if (readEvent != null)
+      changes.add(readEvent)
+    if (writeEvent != null)
+      changes.add(writeEvent)
+
+    callbacks(fd.toLong) = cb
+
+    () => {
```

Comment:
The comment is well placed. It anticipates & answers
the question of a person tracing the code.

Can it be expanded just a bit?  Are you saying that 
at that point the fd has been closed (by somebody), and
that notified kexec to un-register it.

Or are you saying that `callbacks.remove()` will close the fd?
I am now tracing `callbacks`. It looks like a Set and it would
be mildly surprising for a Set.remove to close/transform the thing being removed.

Reply:
Ah yes I see it is confusing as-written. It is the former: we do not need to anything special to unregister the fd with the kqueue, because it will be unregistered automatically when the fd is closed by its owner.

The purpose of `callbacks` is to prevent them from being garbage-collected while the `kqueue` holds the pointer, which the GC doesn't see. I will add a comment to clarify that as well.